From 10f671226d229fffa4c31ac9e068e1c10daf883e Mon Sep 17 00:00:00 2001
From: Andrey Filipenkov <decapitator@ukr.net>
Date: Wed, 2 Nov 2022 13:38:40 +0300
Subject: [PATCH 12/21] fix building on macOS

---
 attr.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/attr.c b/attr.c
index 6253e84..0ce9375 100644
--- a/attr.c
+++ b/attr.c
@@ -18,6 +18,8 @@
 #include <utime.h>
 #include <errno.h>
 #include <dirent.h>
+#include <stdlib.h>
+#include <string.h>
 
 #include "backend.h"
 #include "nfs.h"
-- 
2.42.0

