Upstream-Status: Pending
diff -urp old/bfd/elf-bfd.h new/bfd/elf-bfd.h
--- old/bfd/elf-bfd.h	2020-02-08 13:50:13.000000000 +0100
+++ new/bfd/elf-bfd.h	2020-12-06 15:36:49.000000000 +0100
@@ -26,6 +26,7 @@
 #include "elf/external.h"
 #include "elf/internal.h"
 #include "bfdlink.h"
+#include <string.h>
 
 #ifndef ENABLE_CHECKING
 #define ENABLE_CHECKING 0
diff -urp old/gdbsupport/enum-flags.h new/gdbsupport/enum-flags.h
--- old/gdbsupport/enum-flags.h	2019-09-21 00:04:41
+++ new/gdbsupport/enum-flags.h	2024-04-14 15:53:47
@@ -18,6 +18,8 @@
 #ifndef COMMON_ENUM_FLAGS_H
 #define COMMON_ENUM_FLAGS_H
 
+#include "diagnostics.h"
+
 #include "traits.h"
 
 /* Type-safe wrapper for enum flags.  enum flags are enums where the
diff -urp old/include/diagnostics.h new/include/diagnostics.h
--- old/include/diagnostics.h	2019-02-27 05:51:50
+++ new/include/diagnostics.h	2024-04-14 15:55:30
@@ -63,6 +63,11 @@
 # define DIAGNOSTIC_IGNORE_FORMAT_NONLITERAL \
   DIAGNOSTIC_IGNORE ("-Wformat-nonliteral")
 
+# if __has_warning ("-Wenum-constexpr-conversion")
+#  define DIAGNOSTIC_IGNORE_ENUM_CONSTEXPR_CONVERSION \
+   DIAGNOSTIC_IGNORE ("-Wenum-constexpr-conversion")
+# endif
+
 # if __has_warning ("-Wuser-defined-warnings")
 #  define DIAGNOSTIC_IGNORE_USER_DEFINED_WARNINGS \
    DIAGNOSTIC_IGNORE ("-Wuser-defined-warnings")
@@ -102,6 +107,10 @@
 #ifndef DIAGNOSTIC_IGNORE_FORMAT_NONLITERAL
 # define DIAGNOSTIC_IGNORE_FORMAT_NONLITERAL
 #endif
+
+#ifndef DIAGNOSTIC_IGNORE_ENUM_CONSTEXPR_CONVERSION
+# define DIAGNOSTIC_IGNORE_ENUM_CONSTEXPR_CONVERSION
+#endif
 
 #ifndef DIAGNOSTIC_IGNORE_USER_DEFINED_WARNINGS
 # define DIAGNOSTIC_IGNORE_USER_DEFINED_WARNINGS
diff -urp old/zlib/zutil.h new/zlib/zutil.h
--- old/zlib/zutil.h	2025-04-16 20:58:51
+++ new/zlib/zutil.h	2025-04-16 21:50:21
@@ -137,17 +137,8 @@ extern z_const char * const z_errmsg[10]; /* indexed b
 #  endif
 #endif
 
-#if defined(MACOS) || defined(TARGET_OS_MAC)
+#if defined(MACOS)
 #  define OS_CODE  7
-#  ifndef Z_SOLO
-#    if defined(__MWERKS__) && __dest_os != __be_os && __dest_os != __win32_os
-#      include <unix.h> /* for fdopen */
-#    else
-#      ifndef fdopen
-#        define fdopen(fd,mode) NULL /* No fdopen() */
-#      endif
-#    endif
-#  endif
 #endif
 
 #ifdef __acorn
