diff --git a/src/include/env.h b/src/include/env.h
index b8f7a403a..bb72ce8ca 100644
--- a/src/include/env.h
+++ b/src/include/env.h
@@ -32,8 +32,8 @@
 #define BOX64_NOBANNER_DEFAULT (0)
 #else
 extern char* ftrace_name;
-#define DEFAULT_LOG_LEVEL (ftrace_name ? LOG_INFO : (isatty(fileno(stdout)) ? LOG_INFO : LOG_NONE))
-#define BOX64_NOBANNER_DEFAULT (isatty(fileno(stdout)) ? 0 : 1)
+#define DEFAULT_LOG_LEVEL (LOG_NONE)
+#define BOX64_NOBANNER_DEFAULT (1)
 #endif
 
 #define ENVSUPER1()                                                           \
