diff -urp old/src/include/env.h new/src/include/env.h
--- old/src/include/env.h	2025-02-15 13:38:48
+++ new/src/include/env.h	2025-02-15 13:43:15
@@ -78,10 +78,10 @@ extern char* ftrace_name;
     BOOLEAN(BOX64_LIBCEF, libcef, 1)                                    \
     STRING(BOX64_LIBGL, libgl)                                          \
     ADDRESS(BOX64_LOAD_ADDR, load_addr)                                 \
-    INTEGER(BOX64_LOG, log, DEFAULT_LOG_LEVEL, 0, 3)                    \
+    INTEGER(BOX64_LOG, log, LOG_NONE, 0, 3)                             \
     INTEGER(BOX64_MALLOC_HACK, malloc_hack, 0, 0, 2)                    \
     INTEGER(BOX64_MAXCPU, new_maxcpu, 0, 0, 100)                        \
-    BOOLEAN(BOX64_NOBANNER, nobanner, (isatty(fileno(stdout)) ? 0 : 1)) \
+    BOOLEAN(BOX64_NOBANNER, nobanner, 1)                                \
     STRING(BOX64_NODYNAREC, nodynarec)                                  \
     BOOLEAN(BOX64_NOGTK, nogtk, 0)                                      \
     BOOLEAN(BOX64_NOPULSE, nopulse, 0)                                  \
