diff -urp old/src/core.c new/src/core.c
--- old/src/core.c	2024-11-13 20:14:56
+++ new/src/core.c	2024-11-13 20:19:11
@@ -644,6 +644,7 @@ void LoadLogEnv()
                 box64_nobanner = p[0]-'0';
         }
     }
+    box64_nobanner = 1;
     // grab BOX64_TRACE_FILE envvar, and change %pid to actual pid is present in the name
     openFTrace(NULL, 0);
     box64_log = ftrace_name?LOG_INFO:(isatty(fileno(stdout))?LOG_INFO:LOG_NONE); //default LOG value different if stdout is redirected or not
@@ -672,6 +673,7 @@ void LoadLogEnv()
         if(!box64_nobanner)
             printf_log(LOG_INFO, "Debug level is %d\n", box64_log);
     }
+    box64_log = LOG_NONE;
     if((box64_nobanner || box64_log) && ftrace==stdout)
         box64_stdout_no_w = 1;
 
