diff --git a/src/main.c b/src/main.c
index 5f73ab4bb..127ebffc1 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1,9 +1,57 @@
 #include "core.h"
+#include "debug.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/prctl.h>
+
+static void setupTSOEmulation(void) {
+  int result, box64_hw_tso = 1;
+  const char *p;
+#ifndef PR_GET_MEM_MODEL
+#define PR_GET_MEM_MODEL 0x6d4d444c
+#endif
+#ifndef PR_SET_MEM_MODEL
+#define PR_SET_MEM_MODEL 0x4d4d444c
+#endif
+#ifndef PR_SET_MEM_MODEL_DEFAULT
+#define PR_SET_MEM_MODEL_DEFAULT 0
+#endif
+#ifndef PR_SET_MEM_MODEL_TSO
+#define PR_SET_MEM_MODEL_TSO 1
+#endif
+  // Check to see if TSO is supported.
+  result = prctl(PR_GET_MEM_MODEL, 0, 0, 0, 0);
+  if(result == -1) {
+    perror("TSO not supported");
+    // Unsupported, early exit.
+    return;
+  }
+
+  p = getenv("BOX64_HW_TSO");
+  if(p) {
+    if(strlen(p)==1) {
+      if(p[0]>='0' && p[0]<='0'+1)
+        box64_hw_tso = p[0]-'0';
+    }
+  }
+  if(box64_hw_tso == 0)
+    return;
+
+  // Try to set the TSO mode if we are currently default.
+  result = prctl(PR_SET_MEM_MODEL, PR_SET_MEM_MODEL_TSO, 0, 0, 0);
+  if (result != 0) {
+    perror("TSO not enabled");
+  }
+}
 
 int main(int argc, const char **argv, char **env) {
 
     x64emu_t* emu = NULL;
     elfheader_t* elf_header = NULL;
+
+    setupTSOEmulation();
+
     if (initialize(argc, argv, env, &emu, &elf_header, 1)) {
         return -1;
     }
