diff -ur old/llvm/lib/Support/LockFileManager.cpp new/llvm/lib/Support/LockFileManager.cpp
--- old/llvm/lib/Support/LockFileManager.cpp
+++ new/llvm/lib/Support/LockFileManager.cpp
@@ -36,7 +36,7 @@
 #endif
 
 #if defined(__APPLE__) && defined(__MAC_OS_X_VERSION_MIN_REQUIRED) && (__MAC_OS_X_VERSION_MIN_REQUIRED > 1050)
-#define USE_OSX_GETHOSTUUID 1
+#define USE_OSX_GETHOSTUUID 0
 #else
 #define USE_OSX_GETHOSTUUID 0
 #endif
diff -urp old/llvm/tools/dsymutil/SymbolMap.cpp new/llvm/tools/dsymutil/SymbolMap.cpp
--- old/llvm/tools/dsymutil/SymbolMap.cpp	2022-01-28 10:02:49
+++ new/llvm/tools/dsymutil/SymbolMap.cpp	2023-12-25 12:56:58
@@ -14,7 +14,7 @@
 #include "llvm/Support/Path.h"
 #include "llvm/Support/WithColor.h"
 
-#ifdef __APPLE__
+#if 0//def __APPLE__
 #include <CoreFoundation/CoreFoundation.h>
 #include <uuid/uuid.h>
 #endif
@@ -66,7 +66,7 @@ SymbolMapTranslator SymbolMapLoader::Load(StringRef In
 
   std::string SymbolMapPath = SymbolMap;
 
-#if __APPLE__
+#if 0//__APPLE__
   // Look through the UUID Map.
   if (sys::fs::is_directory(SymbolMapPath) && !Map.getUUID().empty()) {
     uuid_string_t UUIDString;
