diff --git a/arch/arm/boot/dts/omap4.dtsi b/arch/arm/boot/dts/omap4.dtsi
index 96531285..ce560133 100644
--- a/arch/arm/boot/dts/omap4.dtsi
+++ b/arch/arm/boot/dts/omap4.dtsi
@@ -413,6 +413,16 @@
 			#hwlock-cells = <1>;
 		};
 
+		sgx: sgx@0x56000000 {
+			compatible = "ti,omap4-sgx540-120";
+			reg = <0x5600fe00 0x200>;
+			reg-names = "gpu_wrapper";
+			interrupts = <GIC_SPI 21 IRQ_TYPE_LEVEL_HIGH>;
+			ti,hwmods = "gpu";
+			clocks = <&l3_div_ck>, <&sgx_clk_mux>;
+			clock-names = "gpu_iclk", "gpu_fclk";
+		};
+
 		i2c1: i2c@48070000 {
 			compatible = "ti,omap4-i2c";
 			reg = <0x48070000 0x100>;
