diff --git a/arch/arm/boot/dts/omap5-igep0050.dts b/arch/arm/boot/dts/omap5-igep0050.dts
index 8fc19218..a71bcf54 100644
--- a/arch/arm/boot/dts/omap5-igep0050.dts
+++ b/arch/arm/boot/dts/omap5-igep0050.dts
@@ -55,6 +55,16 @@
 	vdda-supply = <&ldo7_reg>;
 };
 
+&i2c1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pins>;
+
+	eeprom: eeprom@50 {
+		compatible = "at,24c256";
+		reg = <0x50>;
+	};
+};
+
 &i2c4 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&i2c4_pins>;
@@ -84,6 +94,20 @@
 };
 
 &omap5_pmx_core {
+	i2c1_pins: pinmux_i2c1_pins {
+		pinctrl-single,pins = <
+			0x1b2 (PIN_INPUT | MUX_MODE0)			/* i2c1_scl */
+			0x1b4 (PIN_INPUT | MUX_MODE0)			/* i2c1_sda */
+		>;
+	};
+
+	i2c2_pins: pinmux_i2c2_pins {
+		pinctrl-single,pins = <
+			0x1f8 (PIN_INPUT | MUX_MODE0)			/* i2c2_scl */
+			0x1fa (PIN_INPUT | MUX_MODE0)			/* i2c2_sda */
+		>;
+	};
+
 	i2c4_pins: pinmux_i2c4_pins {
 		pinctrl-single,pins = <
 			OMAP5_IOPAD(0x0f8, PIN_INPUT | MUX_MODE0)	/* i2c4_scl */
@@ -91,6 +115,18 @@
 		>;
 	};
 
+	mmc1_pins: pinmux_mmc1_pins {
+		pinctrl-single,pins = <
+			0x1a2	0x118		/* sdcard_clk INPUT | MODE0 */
+			0x1a4	0x118		/* sdcard_cmd INPUT | MODE0 */
+			0x1a6	0x118		/* sdcard_data2 INPUT | MODE0 */
+			0x1a8	0x118		/* sdcard_data3 INPUT | MODE0 */
+			0x1aa	0x118		/* sdcard_data0 INPUT | MODE0 */
+			0x1ac	0x118		/* sdcard_data1 INPUT | MODE0 */
+			0x194	0x11e		/* UART6_RTS INPUT | PULLUP  | MODE 6 | GPIO 152 */
+		>;
+	};
+
 	power_button_pin: pinctrl_power_button_pin {
 		pinctrl-single,pins = <
 			OMAP5_IOPAD(0x086, PIN_INPUT | MUX_MODE6)	/* gpio4_118 */
@@ -116,3 +152,26 @@
 		OMAP5_IOPAD(0x1ca, PIN_OUTPUT | MUX_MODE6)	/* perslimbus2_clock.gpio5_145 */
 	>;
 };
+
+&mmc1 {
+	status = "okay";
+	vmmc-supply = <&ldo9_reg>;
+	cd-gpios = <&gpio5 0 GPIO_ACTIVE_LOW>;
+};
+
+&usbhsehci {
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	hub@2 {
+		compatible = "usb424,3503";
+		reg = <2>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ethernet: usbether@3 {
+			compatible = "usb424,7500";
+			reg = <3>;
+		};
+	};
+};
