diff --git a/drivers/gpu/drm/omapdrm/omap_plane.c b/drivers/gpu/drm/omapdrm/omap_plane.c
index 66ac8c40..c90316dc 100644
--- a/drivers/gpu/drm/omapdrm/omap_plane.c
+++ b/drivers/gpu/drm/omapdrm/omap_plane.c
@@ -151,7 +151,7 @@ static void omap_plane_atomic_disable(struct drm_plane *plane,
 
 	plane->state->rotation = DRM_ROTATE_0;
 	omap_state->zorder = plane->type == DRM_PLANE_TYPE_PRIMARY
-			   ? 0 : omap_plane->id;
+			   ? 3 : 3 - omap_plane->id;
 
 	dispc_ovl_enable(omap_plane->id, false);
 }
@@ -220,7 +220,7 @@ void omap_plane_install_properties(struct drm_plane *plane,
 		drm_object_attach_property(obj, prop, 0);
 	}
 
-	drm_object_attach_property(obj, priv->zorder_prop, 0);
+	drm_object_attach_property(obj, priv->zorder_prop, 3);
 }
 
 static struct drm_plane_state *
@@ -268,7 +268,7 @@ static void omap_plane_reset(struct drm_plane *plane)
 	 * plane.
 	 */
 	omap_state->zorder = plane->type == DRM_PLANE_TYPE_PRIMARY
-			   ? 0 : omap_plane->id;
+			   ? 3 : 3 - omap_plane->id;
 	omap_state->base.rotation = DRM_ROTATE_0;
 
 	plane->state = &omap_state->base;
