diff -urp old/video/out/drm_common.c new/video/out/drm_common.c
--- old/video/out/drm_common.c	2023-06-09 23:17:38
+++ new/video/out/drm_common.c	2023-06-09 23:22:12
@@ -999,6 +999,7 @@ bool vo_drm_init(struct vo *vo)
         .card_no = -1,
     };
 
+#if HAVE_X11
     drm->vt_switcher_active = vt_switcher_init(&drm->vt_switcher, drm->log);
     if (drm->vt_switcher_active) {
         vt_switcher_acquire(&drm->vt_switcher, acquire_vt, drm);
@@ -1006,6 +1007,7 @@ bool vo_drm_init(struct vo *vo)
     } else {
         MP_WARN(drm, "Failed to set up VT switcher. Terminal switching will be unavailable.\n");
     }
+#endif
 
     drm->opts = mp_get_config_group(drm, drm->vo->global, &drm_conf);
 
