diff -urp old/kernel/irq/irqdesc.c new/kernel/irq/irqdesc.c
--- old/kernel/irq/irqdesc.c	2015-01-10 14:16:59.000000000 +0100
+++ new/kernel/irq/irqdesc.c	2015-01-10 14:17:10.000000000 +0100
@@ -365,7 +365,7 @@ int __handle_domain_irq(struct irq_domai
 	 * Some hardware gives randomly wrong interrupts.  Rather
 	 * than crashing, do something sensible.
 	 */
-	if (unlikely(!irq || irq >= nr_irqs)) {
+	if (unlikely(irq >= nr_irqs)) {
 		ack_bad_irq(irq);
 		ret = -EINVAL;
 	} else {
