Upstream-Status: Pending
diff -urp old/lib/rpmscript.c new/lib/rpmscript.c
--- old/lib/rpmscript.c	2020-11-10 08:05:56.000000000 +0100
+++ new/lib/rpmscript.c	2020-11-10 10:42:04.000000000 +0100
@@ -14,6 +14,7 @@
 #include "rpmio/rpmlua.h"
 #include "lib/rpmscript.h"
 #include "rpmio/rpmio_internal.h"
+#include <signal.h>
 
 #include "lib/rpmplugins.h"     /* rpm plugins hooks */
 #include "lib/rpmchroot.h"      /* rpmChrootOut */
diff -urp old/lib/poptALL.c new/lib/poptALL.c
--- old/lib/poptALL.c	2020-11-11 22:01:27.000000000 +0100
+++ new/lib/poptALL.c	2020-11-11 22:03:28.000000000 +0100
@@ -46,7 +46,7 @@ rpmQueryFlags rpmcliQueryFlags;
 
 rpmVSFlags rpmcliVSFlags;
 
-int rpmcliVfyLevelMask;
+int rpmcliVfyLevelMask = 0;
 
 extern int _rpmio_debug;
 
diff -urp old/lib/poptQV.c new/lib/poptQV.c
--- old/lib/poptQV.c	2020-11-11 22:01:27.000000000 +0100
+++ new/lib/poptQV.c	2020-11-11 21:58:34.000000000 +0100
@@ -10,7 +10,7 @@
 
 #include "debug.h"
 
-struct rpmQVKArguments_s rpmQVKArgs;
+struct rpmQVKArguments_s rpmQVKArgs = {};
 
 #define POPT_QUERYFORMAT	-1000
 #define POPT_WHATREQUIRES	-1001
