diff --git a/mm/vmalloc.c b/mm/vmalloc.c
index 4165304d3..e61ee27c2 100644
--- a/mm/vmalloc.c
+++ b/mm/vmalloc.c
@@ -3131,6 +3131,7 @@ void *__vmalloc_node_range(unsigned long size, unsigned long align,
 
 	return NULL;
 }
+EXPORT_SYMBOL(__vmalloc_node_range);
 
 /**
  * __vmalloc_node - allocate virtually contiguous memory
diff -ur old/mm/pgtable-generic.c new/mm/pgtable-generic.c
--- old/mm/pgtable-generic.c	2023-08-27 23:49:51
+++ new/mm/pgtable-generic.c	2023-11-19 21:39:49
@@ -285,3 +285,4 @@
 	pte_unmap_unlock(pte, ptl);
 	goto again;
 }
+EXPORT_SYMBOL(__pte_offset_map_lock);
