diff --git a/arch/arm/boot/dts/am57xx-beagle-x15-common.dtsi b/arch/arm/boot/dts/am57xx-beagle-x15-common.dtsi
index fa693289d..966d75eb7 100644
--- a/arch/arm/boot/dts/am57xx-beagle-x15-common.dtsi
+++ b/arch/arm/boot/dts/am57xx-beagle-x15-common.dtsi
@@ -618,7 +618,7 @@
 	vdda_video-supply = <&ldoln_reg>;
 };
 
-&bb2d {
+&gpu_2d {
 	status = "okay";
 };
 
diff --git a/arch/arm/boot/dts/dra7.dtsi b/arch/arm/boot/dts/dra7.dtsi
index 85b60a6ee..a0401ea0b 100644
--- a/arch/arm/boot/dts/dra7.dtsi
+++ b/arch/arm/boot/dts/dra7.dtsi
@@ -1996,8 +1996,8 @@
 			status = "disabled";
 		};
 
-		bb2d: bb2d@59000000 {
-			compatible = "ti,dra7-bb2d";
+		gpu_2d: bb2d@59000000 {
+			compatible = "ti,dra7-bb2d", "vivante,gc";
 			reg = <0x59000000 0x0700>;
 			interrupts = <GIC_SPI 120 IRQ_TYPE_LEVEL_HIGH>;
 			ti,hwmods = "bb2d";
@@ -2212,6 +2212,10 @@
 		#include "dra7-iva-thermal.dtsi"
 	};
 
+	gpu-subsystem {
+		compatible = "ti,dra7-gpu-subsystem";
+		cores = <&gpu_2d>;
+	};
 };
 
 &cpu_thermal {
