--- a/drivers/gpu/drm/omapdrm/omap_plane.c
+++ b/drivers/gpu/drm/omapdrm/omap_plane.c
@@ -98,7 +98,7 @@ static void omap_plane_atomic_disable(struct drm_plane *plane,
 
 	plane->state->rotation = DRM_MODE_ROTATE_0;
 	plane->state->zpos = plane->type == DRM_PLANE_TYPE_PRIMARY
-			   ? 0 : old_omap_state->overlay->overlay_id;
+			   ? 3 : 3 - old_omap_state->overlay->overlay_id;
 
 	omap_overlay_disable(old_state->state, plane, old_omap_state->overlay);
 	new_omap_state->overlay = NULL;
@@ -179,7 +179,7 @@ void omap_plane_install_properties(struct drm_plane *plane,
 						   DRM_MODE_ROTATE_0);
 	}
 
-	drm_object_attach_property(obj, priv->zorder_prop, 0);
+	drm_object_attach_property(obj, priv->zorder_prop, 3);
 }
 
 static void omap_plane_reset(struct drm_plane *plane)
@@ -195,7 +195,7 @@ static void omap_plane_reset(struct drm_plane *plane)
 	 * plane.
 	 */
 	plane->state->zpos = plane->type == DRM_PLANE_TYPE_PRIMARY
-			   ? 0 : omap_plane->id;
+			   ? 3 : 3 - omap_plane->id;
 	plane->state->color_encoding = DRM_COLOR_YCBCR_BT601;
 	plane->state->color_range = DRM_COLOR_YCBCR_FULL_RANGE;
 
