diff --git a/scripts/kconfig/mconf-cfg.sh b/scripts/kconfig/mconf-cfg.sh
index b520e407a..c432286a1 100755
--- a/scripts/kconfig/mconf-cfg.sh
+++ b/scripts/kconfig/mconf-cfg.sh
@@ -1,6 +1,7 @@
 #!/bin/sh
 # SPDX-License-Identifier: GPL-2.0
 
+exit 0
 cflags=$1
 libs=$2
 
diff --git a/scripts/mod/file2alias.c b/scripts/mod/file2alias.c
index 5258247d7..6e4412b2b 100644
--- a/scripts/mod/file2alias.c
+++ b/scripts/mod/file2alias.c
@@ -42,9 +42,6 @@ typedef struct {
 	__u8 b[16];
 } uuid_le;
 
-typedef struct {
-	__u8 b[16];
-} uuid_t;
 
 #define	UUID_STRING_LEN		36
 
@@ -104,17 +101,6 @@ static inline void add_wildcard(char *str)
 		strcat(str + len, "*");
 }
 
-static inline void add_uuid(char *str, uuid_le uuid)
-{
-	int len = strlen(str);
-
-	sprintf(str + len, "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
-		uuid.b[3], uuid.b[2], uuid.b[1], uuid.b[0],
-		uuid.b[5], uuid.b[4], uuid.b[7], uuid.b[6],
-		uuid.b[8], uuid.b[9], uuid.b[10], uuid.b[11],
-		uuid.b[12], uuid.b[13], uuid.b[14], uuid.b[15]);
-}
-
 static inline void add_guid(char *str, guid_t guid)
 {
 	int len = strlen(str);
@@ -1217,24 +1192,6 @@ static int do_cpu_entry(const char *filename, void *symval, char *alias)
 	return 1;
 }
 
-/* Looks like: mei:S:uuid:N:* */
-static int do_mei_entry(const char *filename, void *symval,
-			char *alias)
-{
-	DEF_FIELD_ADDR(symval, mei_cl_device_id, name);
-	DEF_FIELD_ADDR(symval, mei_cl_device_id, uuid);
-	DEF_FIELD(symval, mei_cl_device_id, version);
-
-	sprintf(alias, MEI_CL_MODULE_PREFIX);
-	sprintf(alias + strlen(alias), "%s:",  (*name)[0]  ? *name : "*");
-	add_uuid(alias, *uuid);
-	ADD(alias, ":", version != MEI_CL_VERSION_ANY, version);
-
-	strcat(alias, ":*");
-
-	return 1;
-}
-
 /* Looks like: rapidio:vNdNavNadN */
 static int do_rio_entry(const char *filename,
 			void *symval, char *alias)
@@ -1347,21 +1304,6 @@ static int do_typec_entry(const char *filename, void *symval, char *alias)
 	return 1;
 }
 
-/* Looks like: tee:uuid */
-static int do_tee_entry(const char *filename, void *symval, char *alias)
-{
-	DEF_FIELD_ADDR(symval, tee_client_device_id, uuid);
-
-	sprintf(alias, "tee:%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
-		uuid->b[0], uuid->b[1], uuid->b[2], uuid->b[3], uuid->b[4],
-		uuid->b[5], uuid->b[6], uuid->b[7], uuid->b[8], uuid->b[9],
-		uuid->b[10], uuid->b[11], uuid->b[12], uuid->b[13], uuid->b[14],
-		uuid->b[15]);
-
-	add_wildcard(alias);
-	return 1;
-}
-
 /* Looks like: wmi:guid */
 static int do_wmi_entry(const char *filename, void *symval, char *alias)
 {
@@ -1508,7 +1450,6 @@ static const struct devtable devtable[] = {
 	{"mipscdmm", SIZE_mips_cdmm_device_id, do_mips_cdmm_entry},
 	{"x86cpu", SIZE_x86_cpu_id, do_x86cpu_entry},
 	{"cpu", SIZE_cpu_feature, do_cpu_entry},
-	{"mei", SIZE_mei_cl_device_id, do_mei_entry},
 	{"rapidio", SIZE_rio_device_id, do_rio_entry},
 	{"ulpi", SIZE_ulpi_device_id, do_ulpi_entry},
 	{"hdaudio", SIZE_hda_device_id, do_hda_entry},
@@ -1516,7 +1457,6 @@ static const struct devtable devtable[] = {
 	{"fslmc", SIZE_fsl_mc_device_id, do_fsl_mc_entry},
 	{"tbsvc", SIZE_tb_service_id, do_tbsvc_entry},
 	{"typec", SIZE_typec_device_id, do_typec_entry},
-	{"tee", SIZE_tee_client_device_id, do_tee_entry},
 	{"wmi", SIZE_wmi_device_id, do_wmi_entry},
 	{"mhi", SIZE_mhi_device_id, do_mhi_entry},
 	{"auxiliary", SIZE_auxiliary_device_id, do_auxiliary_entry},
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index b2c0fce3f..48496bc57 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -240,7 +240,6 @@ config X86
 	select HAVE_NOINSTR_HACK		if HAVE_OBJTOOL
 	select HAVE_NMI
 	select HAVE_NOINSTR_VALIDATION		if HAVE_OBJTOOL
-	select HAVE_OBJTOOL			if X86_64
 	select HAVE_OPTPROBES
 	select HAVE_PCSPKR_PLATFORM
 	select HAVE_PERF_EVENTS
diff -urp old/include/linux/objtool.h new/include/linux/objtool.h
--- old/include/linux/objtool.h	2023-11-18 12:54:51
+++ new/include/linux/objtool.h	2023-11-18 12:56:06
@@ -4,7 +4,7 @@
 
 #include <linux/objtool_types.h>
 
-#ifdef CONFIG_OBJTOOL
+#if 1
 
 #include <asm/asm.h>
 
