diff -urp old/Makefile new/Makefile
--- old/Makefile	2013-01-05 20:17:05.000000000 +0100
+++ new/Makefile	2013-01-05 20:26:47.000000000 +0100
@@ -245,7 +245,7 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH"
 
 HOSTCC       = gcc
 HOSTCXX      = g++
-HOSTCFLAGS   = -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer
+HOSTCFLAGS   = -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer $(HOST_INC)
 HOSTCXXFLAGS = -O2
 
 # Decide whether to build built-in, modular, or both.
diff -urp old/scripts/mod/sumversion.c new/scripts/mod/sumversion.c
--- old/scripts/mod/sumversion.c	2013-01-05 20:17:13.000000000 +0100
+++ new/scripts/mod/sumversion.c	2013-01-05 20:28:53.000000000 +0100
@@ -10,6 +10,10 @@
 #include <limits.h>
 #include "modpost.h"
 
+#ifndef PATH_MAX
+#define PATH_MAX 1024
+#endif
+
 /*
  * Stolen form Cryptographic API.
  *
diff -urp old/scripts/kconfig/lxdialog/check-lxdialog.sh new/scripts/kconfig/lxdialog/check-lxdialog.sh
--- old/scripts/kconfig/lxdialog/check-lxdialog.sh	2020-04-24 17:16:52.000000000 +0200
+++ new/scripts/kconfig/lxdialog/check-lxdialog.sh	2020-04-24 17:39:53.000000000 +0200
@@ -75,7 +75,7 @@ case "$1" in
 	"-ldflags")
 		shift
 		cc="$@"
-		ldflags
+		echo '-lncurses'
 		;;
 	"*")
 		usage
