diff -urp old/drivers/mfd/ipaq-micro.c new/drivers/mfd/ipaq-micro.c
--- old/drivers/mfd/ipaq-micro.c	2015-01-01 15:11:29.000000000 +0100
+++ new/drivers/mfd/ipaq-micro.c	2015-01-01 13:10:59.000000000 +0100
@@ -315,7 +315,7 @@ static void micro_reset_comm(struct ipaq
 	rx->state = STATE_SOF;  /* Reset the state machine */
 
 	/* Set up interrupts */
-	writel(0x01, micro->sdlc + 0x0); /* Select UART mode */
+	writel(0x01, micro->sdlc + 0x60); /* Select UART mode */
 
 	/* Clean up CR3 */
 	writel(0x0, micro->base + UTCR3);
