diff -urp old/arch/arm/mach-sa1100/h3xxx.c new/arch/arm/mach-sa1100/h3xxx.c
--- old/arch/arm/mach-sa1100/h3xxx.c	2015-01-01 19:35:19.000000000 +0100
+++ new/arch/arm/mach-sa1100/h3xxx.c	2015-01-01 17:08:32.000000000 +0100
@@ -39,9 +39,13 @@ static struct mtd_partition h3xxx_partit
 		.offset		= 0,
 		.mask_flags	= MTD_WRITEABLE,  /* force read-only */
 	}, {
+		.name		= "H3XXX kernel",
+		.size		= 0x00300000,
+		.offset		= 0x00040000,
+	}, {
 		.name		= "H3XXX rootfs",
 		.size		= MTDPART_SIZ_FULL,
-		.offset		= 0x00040000,
+		.offset		= 0x00340000,
 	}
 };
 
