diff -urp old/rtp.c new/rtp.c
--- old/rtp.c	2024-07-24 22:18:41
+++ new/rtp.c	2024-07-25 06:07:01
@@ -2313,8 +2313,10 @@ void *rtp_buffered_audio_processor(void *arg) {
   // push a deallocator -- av_packet_free(pkt);
   pthread_cleanup_push(swr_alloc_cleanup_handler, &swr);
 
-  av_opt_set_int(swr, "in_channel_layout", AV_CH_LAYOUT_STEREO, 0);
-  av_opt_set_int(swr, "out_channel_layout", AV_CH_LAYOUT_STEREO, 0);
+  AVChannelLayout ch_layout;
+  av_channel_layout_default(&ch_layout, 2);
+  av_opt_set_chlayout(swr, "in_chlayout", &ch_layout, 0);
+  av_opt_set_chlayout(swr, "out_chlayout", &ch_layout, 0);
   av_opt_set_int(swr, "in_sample_rate", conn->input_rate, 0);
   av_opt_set_int(swr, "out_sample_rate", conn->input_rate,
                  0); // must match or the timing will be wrong`
