diff -ur old/scripts/Makefile.build new/scripts/Makefile.build
--- old/scripts/Makefile.build	2022-03-11 10:15:13
+++ new/scripts/Makefile.build	2023-04-29 08:38:01
@@ -574,6 +574,10 @@
 
 # Some files contained in $(targets) are intermediate artifacts.
 # We never want them to be removed automatically.
+ifneq ($(and $(filter notintermediate, $(.FEATURES)),$(filter-out 4.4,$(MAKE_VERSION))),)
+.NOTINTERMEDIATE: $(targets)
+else
 .SECONDARY: $(targets)
+endif
 
 .PHONY: $(PHONY)
