diff -ur old/Makefile.am new/Makefile.am
--- old/Makefile.am	2016-01-31 08:10:16.000000000 +0100
+++ new/Makefile.am	2016-01-31 09:36:07.000000000 +0100
@@ -16,14 +16,10 @@
 #
 
 SUBDIRS = util omx_cam
-bin_PROGRAMS = fliptest v4l2capturedisplay filevpedisplay capturevpedisplay YUVRGBdisplay
-
-if ENABLE_V4L2_DMABUF
-bin_PROGRAMS += dmabuftest
-endif
+bin_PROGRAMS = fliptest
 
 if ENABLE_DCE
-bin_PROGRAMS += viddec3test copycodectest yuvcopytest videnc2test
+bin_PROGRAMS += viddec3test
 endif
 
 LDADD_COMMON = util/libutil.la @DRM_LIBS@ @DCE_LIBS@ @GBM_LIBS@ @EGL_LIBS@ @GLES2_LIBS@
@@ -33,27 +29,7 @@
 fliptest_SOURCES = fliptest.c
 fliptest_LDADD = $(LDADD_COMMON)
 
-v4l2capturedisplay_SOURCES = v4l2capturedisplay.c
-v4l2capturedisplay_LDADD = $(LDADD_COMMON)
-
-filevpedisplay_SOURCES = filevpedisplay.c
-filevpedisplay_LDADD = $(LDADD_COMMON)
-
-capturevpedisplay_SOURCES = capturevpedisplay.c
-capturevpedisplay_LDADD = $(LDADD_COMMON)
-
-if ENABLE_V4L2_DMABUF
-dmabuftest_SOURCES = dmabuftest.c
-dmabuftest_LDADD = $(LDADD_COMMON) -lpthread
-endif
-
 if ENABLE_DCE
 viddec3test_SOURCES = viddec3test.c
 viddec3test_LDADD = $(LDADD_COMMON)
-copycodectest_SOURCES = copycodectest.c
-copycodectest_LDADD = $(LDADD_COMMON)
-yuvcopytest_SOURCES = yuvcopytest.c
-yuvcopytest_LDADD = $(LDADD_COMMON)
-videnc2test_SOURCES = videnc2test.c
-videnc2test_LDADD = $(LDADD_COMMON)
 YUVRGBdisplay_SOURCES = YUVRGBdisplay.c
diff -ur old/configure.ac new/configure.ac
--- old/configure.ac	2016-01-31 08:10:16.000000000 +0100
+++ new/configure.ac	2016-01-31 08:28:54.000000000 +0100
@@ -35,20 +35,6 @@
 # Obtain compiler/linker options for depedencies
 PKG_CHECK_MODULES(DRM, libdrm libdrm_omap)
 
-# Check for kernel headers
-AC_CHECK_HEADER([linux/videodev2.h],
-	[HAVE_V4L2_DMABUF=yes],
-	[HAVE_V4L2_DMABUF=no],
-	[[#include <linux/videodev2.h>]])
-
-if test "x$HAVE_V4L2_DMABUF" = "xyes"; then
-	AC_DEFINE(HAVE_V4L2_DMABUF, 1, [Have V4L2 DMABUF support])
-	AC_MSG_NOTICE([V4L2 DMABUF support detected, building V4L2 DMABUF tests])
-else
-	AC_MSG_WARN([No V4L2 DMABUF support detected, disabling V4L2 DMABUF tests])
-fi
-AM_CONDITIONAL(ENABLE_V4L2_DMABUF, [test "x$HAVE_V4L2_DMABUF" = xyes])
-
 # Check optional KMSCUBE:
 AC_ARG_ENABLE([kmscube], AS_HELP_STRING([--disable-kmscube], [disable kmscube display support]))
 AS_IF([test "x$enable_kmscube" != "xno"], [PKG_CHECK_EXISTS(gbm egl glesv2, [HAVE_KMSCUBE=yes], [HAVE_KMSCUBE=no])])
diff -ur old/util/Makefile.am new/util/Makefile.am
--- old/util/Makefile.am	2016-01-31 08:10:16.000000000 +0100
+++ new/util/Makefile.am	2016-01-31 08:32:27.000000000 +0100
@@ -19,12 +19,7 @@
 
 libutil_la_SOURCES = \
 	display-kms.c \
-	util.c \
-	vpe-common.c
-
-if ENABLE_V4L2_DMABUF
-libutil_la_SOURCES += v4l2.c
-endif
+	util.c
 
 if ENABLE_DCE
 libutil_la_SOURCES += demux.c
