diff -urp old/util/display-kmscube.c new/util/display-kmscube.c
--- old/util/display-kmscube.c	2020-04-02 01:49:25.000000000 +0200
+++ new/util/display-kmscube.c	2020-04-02 01:52:29.000000000 +0200
@@ -1102,7 +1102,7 @@ disp_kmscube_usage(void)
 	MSG("\t--distance <float>\tset cube distance (default 8.0)");
 	MSG("\t--fov <float>\tset field of vision (default 45.0)");
 	MSG("\t--kmscube\tEnable display kmscube (default: disabled)");
-	MSG("\t--connector <connector_id>\tset the connector ID (default: LCD)");
+	MSG("\t--connector <connector_id>\tset the connector ID (default: HDMI)");
 }
 
 struct display *
@@ -1111,7 +1111,7 @@ disp_kmscube_open(int argc, char **argv)
 	struct display_kmscube *disp_kmsc = NULL;
 	struct display *disp;
 	int ret, i, enabled = 0;
-	float fov = 45, distance = 8, connector_id = 4;
+	float fov = 45, distance = 8, connector_id = 7;
 
 	/* note: set args to NULL after we've parsed them so other modules know
 	 * that it is already parsed (since the arg parsing is decentralized)
diff -urp old/util/display-kms.c new/util/display-kms.c
--- old/util/display-kms.c	2020-03-31 22:59:37.000000000 +0200
+++ new/util/display-kms.c	2020-04-02 18:01:35.000000000 +0200
@@ -610,6 +610,11 @@ disp_kms_open(int argc, char **argv)
 
 	disp->multiplanar = true;
 
+	struct connector *c = &disp_kms->connector[disp_kms->connectors_count++];
+	c->id = 7;
+	c->crtc = -1;
+	strcpy(c->mode_str, "1920x1080");
+
 	/* note: set args to NULL after we've parsed them so other modules know
 	 * that it is already parsed (since the arg parsing is decentralized)
 	 */
@@ -675,8 +675,8 @@ disp_kms_open(int argc, char **argv)
 		argv[i] = NULL;
 	}
 
-	disp->width = 0;
-	disp->height = 0;
+	disp->width = 1920;
+	disp->height = 1080;
 	for (i = 0; i < (int)disp_kms->connectors_count; i++) {
 		struct connector *c = &disp_kms->connector[i];
 		connector_find_mode(disp, c);
