diff -urp old/arch/arm/mach-k3/config.mk new/arch/arm/mach-k3/config.mk
--- old/arch/arm/mach-k3/config.mk	2023-12-27 18:56:53
+++ new/arch/arm/mach-k3/config.mk	2023-12-27 19:00:53
@@ -52,7 +52,7 @@ tiboot3.bin: image_check FORCE
 	fi
 
 tiboot3.bin: image_check FORCE
-	$(srctree)/tools/k3_gen_x509_cert.sh -c 16 -b $(obj)/u-boot-spl.bin \
+	bash $(srctree)/tools/k3_gen_x509_cert.sh -c 16 -b $(obj)/u-boot-spl.bin \
 				-o $@ -l $(CONFIG_SPL_TEXT_BASE) -r $(SWRV) -k $(KEY)
 
 INPUTS-y	+= tiboot3.bin
@@ -84,7 +84,7 @@ cmd_k3_mkits = \
 
 quiet_cmd_k3_mkits = MKITS   $@
 cmd_k3_mkits = \
-	$(srctree)/tools/k3_fit_atf.sh \
+	bash $(srctree)/tools/k3_fit_atf.sh \
 	$(CONFIG_K3_ATF_LOAD_ADDR) \
 	$(patsubst %,$(obj)/dts/%.dtb,$(subst ",,$(LIST_OF_DTB))) > $@
 
